SI-7898 Read user input during REPL warmup

som-snytt opened the pull request:
The compiler is created on main thread and user input is read
on an aux thread (opposite to currently).

Fixes completion when `-i` is supplied.

Adds an `-I` option to paste init files.

The temporary reader uses postInit to swap in the underlying
reader.

Completion is disabled for the temporary reader, rather than
blocking while it waits for a compiler. But manically hitting
tab is one way of knowing exactly when completion is live.

som-snytt committed:
SI-7898 Read user input during REPL warmup

The compiler is created on main thread and user input is read
on an aux thread (opposite to currently).

Fixes completion when `-i` is supplied.

Now `-i` means pasted and new option `-I` means line-by-line.

The temporary reader uses postInit to swap in the underlying
reader.

Completion is disabled for the temporary reader, rather than
blocking while it waits for a compiler. But manically hitting
tab is one way of knowing exactly when completion is live.

som-snytt commented:
@SethTisue `-i` is so useful, I wonder if this is worth squeezing into 2.11.8 in order not to cripple the great new completion feature.

Sry for eleventh hour; this is code from a few years ago, etc.

som-snytt commented:
Apparently, I added a test once that shows `-i` output. The other note is that the `-i` noise is suppressed. You can use `-Dscala.repl.debug=true` to see errors in init scripts.

SethTisue commented:
/rebuild

SethTisue commented:
I've opened https://github.com/scala/scala-dev/issues/96 on the (unrelated) integrate-ide failure

SethTisue commented:
/rebuild

SethTisue commented:
> I wonder if this is worth squeezing into 2.11.8

I don't we think we can risk it :-\

som-snytt commented:
It would be nice is this got merged forward to 2.12.

som-snytt commented:
If this got merged, I could use REPL for my tool at work, for which I want some joe to type something in and use autocomplete, but I need either -i for set-up or go totally custom.

lrytz commented on src/compiler/scala/tools/nsc/GenericRunnerCommand.scala:
```diff
@@ -79,6 +79,7 @@ Other startup options:
 
  -howtorun    what to run <script|object|jar|guess> (default: guess)
  -i <file>    preload <file> before starting the repl
+ -I <file>    preload <file> as though pasted
```
to an ordinary user, it's probably not clear what `preload <file> as though pasted` means, and how it behaves differently from `-i` - maybe add something like "i.e., interpreting all at once instead of line-by-line"? you can probably come up with something better..

som-snytt replied:
Reversed the sense. -I is preload line-by-line.

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -175,10 +175,18 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+    if (isReplDebug) op
+    else intp beSilentDuring {
+      mum = true
+      try op finally mum = false
```
you could save the current state of `mum` in a local and finally restore its original value, so nested calls of `mumly` calls would also work

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -408,11 +416,13 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+    line match {
       case null => EOF
-      case line => if (try processLine(line) catch crashRecovery) loop() else ERR
+      case _    => if (try processLine(line) catch crashRecovery) loop(readOneLine()) else ERR
```
i think an `if else if else` would look cleaner here

som-snytt replied:
I'd become if-averse.

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -885,49 +895,104 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+      intp.quietBind(NamedParam[IMain]("$intp", intp)(tagOfIMain, classTag[IMain]))
+
+      // add a help function for anyone who types "help" instead of ":help". Easily shadowed.
+      //addHelp()
```
leftover from some other patch?

som-snytt replied:
the whole PR is leftover.

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -885,49 +895,104 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+          flatMap (f => File(f).safeSlurp())
+          foreach (intp quietRun _)
+      )
+      // classloader and power mode setup
```
stale comment about "classloader setup"? or did you remove `intp.setContextClassLoader()` accidentally?

som-snytt replied:
Removal was https://github.com/scala/scala/pull/4841/files#diff-f31fa4d21a39bc6ed3c30fd38bcbc9d4L892 where the comment was correct, of course.

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -885,49 +895,104 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+      in = sr
+      SplashLoop(sr, prompt)
     }
-    loadFiles(settings)
```
looks like method `loadFiles` is now unused and can be removed

lrytz commented:
the idea to preload files in a separate thread is nice, but there's a drawback of no longer emitting any errors that stem from the loaded file. in current scala

```
$ scala -i Test.scala
Loading Test.scala...
<console>:11: error: not found: value bluberish
       val x = bluberish
               ^

Welcome to Scala 2.12.0-20160502-155348-d550006 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_77).
Type in expressions for evaluation. Or try :help.

scala>
```

after this change:

```
$ ~/scala/scala3/build-sbt/quick/bin/scala -i Test.scala
Welcome to Scala 2.11.8-20160225-221846-903998c (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_77).
Type in expressions for evaluation. Or try :help.

scala> x
<console>:12: error: not found: value x
       x
       ^
```

thoughts? we could put background loading behind another flag.

lrytz commented on src/repl/scala/tools/nsc/interpreter/ILoop.scala:
```diff
@@ -885,49 +895,104 @@ class ILoop(in0: Option[BufferedReader], protected val out: JPrintWriter)
+      case _ =>
+    }
+    // TODO: wait until after startup to enable obnoxious settings
+    def withSuppressedSettings[A](body: =>A): A = {
```
not sure what the future intensions are, but maybe remove that method until they are actually implemented?

som-snytt replied:
I'll check that out. It was to delay setting `-print:typer` for example.

lrytz commented:
another question: would we break anything if we just changed `-i` to paste mode?

som-snytt commented:
I commented earlier, "You can use -Dscala.repl.debug=true to see errors in init scripts." But I've forgotten details by now and must respool. The general idea was not to see start-up cruft, including -print output on start-up which isn't relevant to user.

som-snytt commented:
That's a good question about -i. I guess line-by-line means a line can fail independently. Rarely one does want to limp along. Failure could be due to missing jar for something yagni.

som-snytt commented:
@lrytz Pushed feedback fix-up from last week. I realize @lrytz is busy on weekends. I'd wanted to show a colleague a line from this PR, but was foiled b/c hadn't pushed yet. Push early!

som-snytt committed:
SI-7898 Quiet REPL at startup

Enable noisy modes only when interpreting user input.

som-snytt commented:
Adds quiet startup.

lrytz commented:
I really like all the changes here. I'm just still a bit worried about making `-i` quiet by default, I can imagine it causing frustration, so I'd prefer to have that as a separate option. How about adding `-iq` and `-Iq`?

An even better option might be to only report errors and warnings, but skip the repl log output. Not sure how hard that would be to implement.

lrytz commented:
I understand that it would be annoying to get error messages printed while the user already has a prompt at his hand, but then the solution is "simply" to fix the init script not to produce errors / warnings. or we could have the quiet options on top.

som-snytt commented:
Yes, you're right.

som-snytt committed:
SI-7898 Report paste errors improvedly

Use a "label" for errors, so that script names are shown.
Position is still wrong for scripts in REPL.

Initial scripts are run with `ILoop.echo` and results printing
turned off, but reporter still enabled.

som-snytt commented:
/rebuild

lrytz committed:
SI-7898 Label for parsing -i sources

Text-based REPL pre-parses, so use the current label for errors.

lrytz commented:
i was still getting `<console>` for errors in the `-i` file, here's what made it work: https://github.com/som-snytt/scala/compare/issue/7898-i...lrytz:pr4998?expand=1 - that change is without understanding anything from the control flow, just grepp'd a bit around

som-snytt commented:
@lrytz Apparently, I didn't try a parse error.

lrytz commented:
LGTM!

dwijnand commented:
Cross-linking, this has broken the REPL in sbt: https://issues.scala-lang.org/browse/SI-9804, specifically this commit: https://github.com/scala/scala/commit/99dad60d984d3f72338f3bad4c4fe905090edd51.

som-snytt commented:
@dwijnand Thanks.

Is there an issue somewhere that `publishLocal` should not enforce using java 6?

```
java.lang.RuntimeException: javac version 1.8.0_60 may not be used to publish, it has to start with 1.6 due to javacVersionPrefix setting
```

dwijnand commented:
sbt must work on Java 6, so officially that's required, but you're free to drop that in sbt: https://github.com/sbt/sbt/blob/642b091bf68c3f4f770ae043524c08491881074d/project/Release.scala#L23

But I don't think the fix should be in sbt - can't just break the repl for 2.12.0-M5+ for versions of sbt < whatever version ships with a fix (eg. 0.13.12).

som-snytt commented:
I was building locally because usually some debug is required; but I've looked at the console before, because I remember that initialCommands run very early.

https://github.com/sbt/sbt-javaversioncheck/issues/2
